src/frontend/Deprecation_analysis.ml

```diff
@@ -0,0 +1,163 @@
+      |> (ignore : [`Ok | `Duplicate] -> unit)
+  | _ -> ()
+
+let rec warn_deprecated_expr (acc : (Location_span.t * string) list)
+    ({expr; emeta} : (typed_expr_meta, fun_kind) expr_with) :
+    (Location_span.t * string) list =
+  match expr with
+  | GetLP ->
+      acc
+      @ [ ( emeta.loc
+          , "The no-argument function `get_lp()` is deprecated. Use the \
+             no-argument function `target()` instead." ) ]
+  | FunApp (StanLib, {name= "abs"; _}, [e])
+    when Middle.UnsizedType.is_real_type e.emeta.type_ ->
+      warn_deprecated_expr
+        ( acc
+        @ [ ( emeta.loc
+            , "Use of the `abs` function with real-valued arguments is \
+               deprecated; use functions `fabs` instead." ) ] )
+        e
+  | FunApp (StanLib, {name= "if_else"; _}, l) ->
+      acc
+      @ [ ( emeta.loc
+          , "The function `if_else` is deprecated. Use the conditional \
+             operator (x ? y : z) instead." ) ]
+      @ List.concat (List.map l ~f:(fun e -> warn_deprecated_expr [] e))
+  | FunApp (StanLib, {name; _}, l) ->
+      let w =
+        if Option.is_some (String.Map.find deprecated_distributions name) then
+          [ ( emeta.loc
+            , name ^ " is deprecated and will be removed in the future. Use "
+              ^ rename_distribution name ^ " instead." ) ]
+        else if Option.is_some (String.Map.find deprecated_functions name) then
+          [ ( emeta.loc
+            , name ^ " is deprecated and will be removed in the future. Use "
+              ^ rename_function name ^ " instead." ) ]
+        else []
+      in
+      acc @ w
+      @ List.concat (List.map l ~f:(fun e -> warn_deprecated_expr [] e))
+  | FunApp (UserDefined, {name; _}, l) ->
+      let w =
+        let newname = String.Table.find deprecated_userdefined name in
+        if Option.is_some newname then
+          [ ( emeta.loc
+            , "Use of the _log suffix in user defined function " ^ name
+              ^ " is deprecated, use " ^ Option.value_exn newname ^ " instead."
+            ) ]
+        else []
+      in
+      acc @ w
+      @ List.concat (List.map l ~f:(fun e -> warn_deprecated_expr [] e))
+  | _ -> fold_expression warn_deprecated_expr (fun l _ -> l) acc expr
+
+let warn_deprecated_lval acc l =
+  fold_lval_with warn_deprecated_expr (fun x _ -> x) acc l
+
+let rec warn_deprecated_stmt (acc : (Location_span.t * string) list) {stmt; _}
+    : (Location_span.t * string) list =
+  match stmt with
+  | IncrementLogProb e -> warn_deprecated_expr acc e
+  | Assignment {assign_lhs= l; assign_op= ArrowAssign; assign_rhs= e} ->
+      acc @ warn_deprecated_lval [] l @ warn_deprecated_expr [] e
+  | FunDef {body; _} -> warn_deprecated_stmt acc body
+  | _ ->
+      fold_statement warn_deprecated_expr warn_deprecated_stmt
+        warn_deprecated_lval
+        (fun l _ -> l)
+        acc stmt
+
+let emit_warnings (program : typed_program) : unit =
+  String.Table.clear deprecated_userdefined ;
```

Comment:
We should avoid using global mutable state like `deprecetaed_userdefined` unless strictly necessary. Instead accumulate the results of functions into their return values. Global mutable state is a lot less clear and more error prone.

I think this issue predates this PR, though..

Reply:
Yes, the global `deprecated_userdefined` was my fault back when I had no experience with functional programming.

So right now `replace_suffix` fills in `deprecated_userdefined` and then `warn_deprecated_expr` and `replace_deprecated_expr` look things up in that global hashtable.
A better design would be a `replace_suffix` that builds and returns a `String.Set.t` which is then explicitly passed to `warn_deprecated_expr`/`replace_deprecated_expr`. That requires passing it through a few other functions too but nothing too radical.
The code here, at the beginning of `emit_warnings` would be something like
```ocaml
let deprecated_userdefined =
  program.functionblock |> Option.value ~default:[]
  |> List.filter_map ~f:replace_suffix
  |> String.Set.of_list
in
```

By the way, I'm not a fan of the name `replace_suffix` because the function does not _replace_ anything, rather it _finds_ replaceable distribution names.

Reply:
That's a great idea. I can refactor it to something like
```ocaml
let update_suffix name type_ = 
      let open String in
        if is_suffix ~suffix:"_cdf_log" name then drop_suffix name 8 ^ "_lcdf"
        else if is_suffix ~suffix:"_ccdf_log" name then
          drop_suffix name 9 ^ "_lccdf"
        else if Middle.UnsizedType.is_real_type type_ then
          drop_suffix name 4 ^ "_lpdf"
        else drop_suffix name 4 ^ "_lpmf"

let find_udf_suffixes = function
  | { stmt= FunDef {funname= {name; _}; arguments= (_, type_, _) :: _; _}
    ; smeta= _ }
    when String.is_suffix ~suffix:"_log" name ->
      Some (name, type_)
  | _ -> None
```
but then I have list of `(string * UnsizedType.t)` and I can't use `String.Set.of_list`. Is there a way to keep the type information (necessary for `update suffix`) and use a Set? I can think of hacky solutions to this, but is there a nicer way?

Reply:
Ah, right, a Hashtable is not a _Set_ but a _Map_. You need `String.Map.of_alist_exn`.

Reply:
OK, I've implemented this and I think it makes more sense now. The only thing I'm not sure about: `String.Map.of_alist_exn` requires the keys to be unique (hence the `dedup` step). This can potentially create issues if we have different `foo_log` UDFs with different signatures, but this shouldn't be possible, right?

Reply:
UDF overloading is not supported, at least not yet. So it currently is not a problem.

Reply:
(for reference there is also a `multi` series of tables and maps that gives you a map of lists, e.g. `val of_alist_multi : (key * 'a) list -> 'a list t`)

Reply:
@adamhaber nice job on this fix! Looking good to me